craft_parts/infos.py

```diff
@@ -141,48 +153,65 @@ def project_options(self) -> Dict[str, Any]:
             "application_name": self.application_name,
             "arch_triplet": self.arch_triplet,
             "target_arch": self.target_arch,
+            "project_vars_part_name": self._project_vars_part_name,
+            "project_vars": self._project_vars,
         }
 
     def set_project_var(
-        self, name: str, value: str, *, part_name: Optional[str] = None
+        self,
+        name: str,
+        value: str,
+        *,
+        part_name: Optional[str] = None,
     ) -> None:
         """Set the value of a project variable.
 
-        :param name: The custom argument name.
-        :param value: The new custom argument value.
+        Variable values can be set once. Project variables are not intended for
+        logic construction in user scripts, setting it multiple times is likely to
+        be an error.
 
-        :raise ValueError: If there is no custom argument with the given name,
-            or if the variable has already been consumed by the application.
+        :param name: The project variable name.
+        :param value: The new project variable value.
+        :param part_name: If not None, variable setting is restricted to the named part.
+
+        :raise ValueError: If there is no custom argument with the given name.
+        :raise RuntimeError: If a write-once variable is set a second time, or if a
+            part name is specified and the variable is set from a different part.
         """
         self._ensure_valid_variable_name(name)
 
-        if name in self._consumed_project_vars:
-            raise ValueError(
-                f"cannot set variable {name!r}, it has already been consumed"
-            )
+        if self._project_vars[name].updated:
+            raise RuntimeError(f"variable {name!r} can be set only once")
 
         if self._project_vars_part_name in [None, part_name]:
-            self._project_vars[name] = value
+            self._project_vars[name].value = value
+            self._project_vars[name].updated = True
```

Comment:
just an idea, you could eventually set .updated by overriding __setattr__  or __setitem__

Reply:
I considered that, but I tried to avoid using tricks that could make it harder to follow in the future. But we can improve this later.